python/lsst/faro/utils/stellarLocus.py

```diff
@@ -0,0 +1,232 @@
+        yFiducialHW = np.array(paramsOut['yMin'])
+        xFiducialHW = (yFiducialHW - paramsOut["bODR"])/paramsOut["mODR"]
+        ysFitLine = np.array([yMinLine, yMaxLine])
+        xsFitLine = (ysFitLine - paramsOut["bODR"])/paramsOut["mODR"]
+        ysFitLine2 = np.array([yMinLine, yMaxLine])
+        xsFitLine2 = (ysFitLine2 - paramsOut["bODR2"])/paramsOut["mODR2"]
+        yFiducial = np.array(paramsOut['yMin'])
+        xFiducial = (yFiducial - paramsOut["bODR2"])/paramsOut["mODR2"]
+    else:
+        xsFitLineHW = np.array([xMinLine, xMaxLine])
+        ysFitLineHW = paramsOut["mHW"]*xsFitLineHW + paramsOut["bHW"]
+        xFiducialHW = np.array(paramsOut['xMin'])
+        yFiducialHW = paramsOut["mHW"]*xFiducialHW + paramsOut["bHW"]
+        xsFitLine = np.array([xMinLine, xMaxLine])
+        ysFitLine = [paramsOut["mODR"]*xsFitLine[0] + paramsOut["bODR"],
+                     paramsOut["mODR"]*xsFitLine[1] + paramsOut["bODR"]]
+        xsFitLine2 = np.array([xMinLine, xMaxLine])
+        ysFitLine2 = [paramsOut["mODR2"]*xsFitLine2[0] + paramsOut["bODR2"],
+                      paramsOut["mODR2"]*xsFitLine2[1] + paramsOut["bODR2"]]
+        xFiducial = np.array(paramsOut['xMin'])
+        yFiducial = paramsOut["mODR2"]*xFiducial + paramsOut["bODR2"]
+
+    # Calculate the distances to the fit line
+    # Need two points to characterise the lines we want
+    # to get the distances to
+    pt1 = np.array([xFiducial, yFiducial])
+    pt2 = np.array([xsFitLine2[1], ysFitLine2[1]])
+
+    pt1HW = np.array([xFiducialHW, yFiducialHW])
+    pt2HW = np.array([xsFitLine2[1], ysFitLineHW[1]])
+
+    # "p2" is the distance from the fit line, and "p1" is the distance
+    # along the line.
+    p1HW, p2HW = parallelPerpDistance(pt1HW, pt2HW, zip(xs, ys))
+    p1, p2 = parallelPerpDistance(pt1, pt2, zip(xs, ys))
+
+    return np.array(p1), np.array(p2), paramsOut
+
+
+def parallelPerpDistance(pt1, pt2, points):
+    """Calculate the perpendicular distance to a line from a point, and
+       the distance along the line (parallel) from that point to a fiducial.
+    Parameters
+    ----------
+    pt1 : `numpy.ndarray`
+        A point on the line; this will be taken as the zero point for
+        the parallel distance.
+    pt2 : `numpy.ndarray`
+        Another point on the line
+    points : `zip`
+        The points to calculate the distance to
+    Returns
+    -------
+    dists_parallel : `list`
+        The distances along the line from the points to the fiducial point (pt1).
+    dists_perp : `list`
+        The distances from the line to the points. Uses the cross
+        product to work this out.
+    """
+    dists_parallel = []
+    dists_perp = []
+
+    for point in points:
+        point = np.array(point)
+        distToLine = np.cross(pt1 - point, pt2 - point)/np.linalg.norm(pt2 - pt1)
+        distAlongLine = np.sqrt(distToLine**2 + (np.linalg.norm(pt1 - point))**2)
```

Comment:
I think this is wrong and in the paper they use the distance from the point of maximum density rather than a point on the line.

Reply:
I think the distance along the line should be np.sqrt( |pt1 - p|**2 - distToLine**2 ).

Reply:
> I think this is wrong and in the paper they use the distance from the point of maximum density rather than a random point on the line. There is some code in analysis_drp to calculate the point of maximum density that could be adapted.

I see no mention in Ivezic+2004 of p1 being defined relative to a point of maximum density. As with selecting limits on p1, if this is going to be a choice we make, we should (a) decide on how we'll define it, and (b) write it down.

Reply:
> I think the distance along the line should be np.sqrt( |pt1 - p|^2 - distToLine^2 ).

Oops - I think you're right. I'll confirm and fix it.

Reply:
I assume the wPerp code gives the same answer as before? Does that mean this distance calculation was wrong before and that fixing it now will change the answer? Or that it was right before but seemingly has no effect when comparing with the previous answer?

Reply:
> > I think the distance along the line should be np.sqrt( |pt1 - p|^2 - distToLine^2 ).
> 
> Oops - I think you're right. I'll confirm and fix it.

Fixed - `        distAlongLine = np.sqrt((np.linalg.norm(pt1 - point))**2 - distToLine**2)` is what it should have been.

I'll need to check a direct comparison, but we wouldn't necessarily expect the results from the old algorithm and the new wPerp to be identical, because they are calculated in a slightly different way.